### PR TITLE
Feature/lifecycle authstate

### PR DESCRIPTION
**Pull request by CrazyKsv**

### CODE REVIEW

File: lib/ptolemy_server.ex [hunk 1]
```diff
@@ -124,44 +149,99 @@ defmodule Ptolemy.Server do
...
+    %{
+      auth_mode: mode,
+      credentials: creds,
+      opts: opts
+    } = auth
+    opts = opts |> parse_opts
+    case mode do
+      :GCP ->
+        %{gcp_svc_acc: svc} = creds
+        creds = creds |> Map.put(:gcp_svc_acc, svc |> Poison.decode!)
```
Comment (brsmsn) [inline lib/ptolemy_server.ex]: We should remove `Posison.decode!` and replace it with `Jason.decode!`

File: lib/ptolemy_server.ex [hunk 2]
```diff
@@ -111,10 +119,27 @@ defmodule Ptolemy.Server do
...
+  defp parse_opts(opts) when is_list(opts) do
+
+    case opts do
+      [] -> []
+      _ ->
+        svc = Keyword.get(opts, :iap_svc_acc)
+
+        case svc do
+          :reuse -> opts
+          _ -> opts |> Keyword.replace!(:iap_svc_acc, svc |> Poison.decode!)
```
Comment (brsmsn) [inline lib/ptolemy_server.ex]: Same thing here lets change `Poison.decode!` to `Jason.decode!`

File: lib/ptolemy_server.ex [hunk 3]
```diff
@@ -32,16 +32,24 @@ defmodule Ptolemy.Server do
...
-        {:error, _ } -> 
+        {:ok, tokens} ->
+          [
+            vault: vault_tok,
+            iap: iap_tok
+          ] = tokens
+          %{
+            vault: vault_tok,
+            iap: iap_tok
+          }
```
Comment (brsmsn) [inline lib/ptolemy_server.ex]: The return should be in this format: 
```elixir 
[
   {"Authroization", "Bearer US7aA"}, 
   {"X-Vault-Token", "Aasdasd"}
]
```

**Comment by dishonesthips** [comment]
Looks good after adding tests and

**Review by tangsaidi — APPROVED** [review]

**Review by dishonesthips — APPROVED** [review]

**Review by fvumbaca — APPROVED** [review]

**Pull request merged**